include/ofi_hook.h

```diff
@@ -66,19 +68,54 @@ enum ofi_hook_class {
 extern struct fi_ops hook_fid_ops;
 struct fid *hook_to_hfid(const struct fid *fid);
 struct fid_wait *hook_to_hwait(const struct fid_wait *wait);
+struct hook_prov_ctx {
+	struct fi_provider	prov;
+	int 			(*ini_fid[FI_CLASS_MAX])(struct fid *fid);
+	int 			(*fini_fid[FI_CLASS_MAX])(struct fid *fid);
+};
```

Comment:
With just hook_provider, I couldn't come up with a good name for the field in the following struct. We already have a hprov. Do we just call it prov? That would break the convention of naming struct fi_provider field as prov.

```
struct hook_fabric {
        struct fid_fabric       fabric;
        struct fid_fabric       *hfabric;
        enum ofi_hook_class     hclass;
        struct fi_provider      *hprov;
        struct hook_prov_ctx    *prov_ctx;
};
```

Reply:
I think we could just call it prov.  That aligns with fabric/hfabric, etc.  For full symmetry, I guess you could store struct fi_provider instead of hook_provider, but I wouldn't actually do that.  We require your new struct.

Hmm... another option would be to store the ini/fini calls in a separate structure that we reference from struct fi_prov_context.  We could even extend the definition of fi_prov_context with a union that is accessed based on the prov_type.  That might work better if we want to support external hooks, without the external hook provider needing to implement everything.